File: app/src/main/java/com/example/hitchapp/adapters/RidesAdapter.java
package com.example.hitchapp.adapters;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.hitchapp.R;
import com.example.hitchapp.fragments.ProfileFragment;
import com.example.hitchapp.models.Request;
import com.example.hitchapp.models.Ride;
import com.example.hitchapp.models.User;
import com.parse.GetCallback;
import com.parse.ParseException;
import com.parse.ParseFile;
import com.parse.ParseQuery;
import com.parse.ParseUser;
import com.parse.SaveCallback;

import java.text.SimpleDateFormat;
import java.util.List;

public class RidesAdapter extends RecyclerView.Adapter<RidesAdapter.ViewHolder> {

    private static final String TAG = "RidesAdapter";
    private Context context;
    private List<Ride> rides;

    public RidesAdapter(Context context, List<Ride> rides) {
        this.context = context;
        this.rides = rides;
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_ride, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Ride ride = rides.get(position);
        holder.bind(ride);
    }

    @Override
    public int getItemCount() {
        return rides.size();
    }

    // Clean all elements of the recycler
    public void clear() {
        rides.clear();
        notifyDataSetChanged();
    }

    // Add a list of items -- change to type used
    public void addAll(List<Ride> list) {
        rides.addAll(list);
        notifyDataSetChanged();
    }

    public void setAll(List<Ride> list) {
        clear();
        addAll(list);
        notifyDataSetChanged();
    }


    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private TextView tvFirstName;
        private TextView tvLastName;
        private ImageView ivProfilePicture;
        private TextView tvFrom;
        private TextView tvTo;
        private TextView tvDepartureTime;
        private TextView tvDepartureDate;
        private TextView tvPrice;
        protected Button btnRequest;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            tvFirstName = itemView.findViewById(R.id.tvFirstName);
            tvLastName = itemView.findViewById(R.id.tvLastName);
            tvFrom = itemView.findViewById(R.id.tvFrom);
            tvTo = itemView.findViewById(R.id.tvTo);
            tvDepartureDate = itemView.findViewById(R.id.tvDepartureDate);
            tvDepartureTime = itemView.findViewById(R.id.tvDepartureTime);
            tvPrice = itemView.findViewById(R.id.tvPrice);
            ivProfilePicture = itemView.findViewById(R.id.ivProfilePicure);
            btnRequest = itemView.findViewById(R.id.btnRequest);

            // Add this as the itemView's OnClickListener
            itemView.setOnClickListener(this);

            // Listens for driver profile pic clicked
            profilePicListener();

            // Listens for when someone requests to join ride
            btnRequestListener();

        }

        public void bind(Ride ride) {
            SimpleDateFormat DateFor = new SimpleDateFormat("MM/dd/yyyy");

            // Bind the ride data to the view elements
            User user = (User) ride.getDriver();
            tvFirstName.setText(user.getFirstName());
            tvLastName.setText(user.getLastName());
            tvFrom.setText(ride.getFrom());
            tvTo.setText(ride.getTo());
            tvDepartureTime.setText(ride.getDepartureTime());
            tvDepartureDate.setText(DateFor.format(ride.getDepartureDate()));
            tvPrice.setText(String.valueOf(ride.getPrice()));
            if(ParseUser.getCurrentUser().getObjectId().equals(ride.getDriver().getObjectId()))
                btnRequest.setVisibility(View.GONE);
            else{
                btnRequest.setVisibility(View.VISIBLE);
            }

            ParseFile profile = user.getProfilePicture();
            if (profile != null) {
                Glide.with(context)
                        .load(profile.getUrl())
                        .fitCenter()
                        .circleCrop()
                        .into(ivProfilePicture);
            }
        }

        @Override
        public void onClick(View v) {

        }

        // When someone's profile pic gets clicked you get taken to their profile
        private void profilePicListener() {
            ivProfilePicture.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int position = getAdapterPosition();
                    // Make sure the position is valid i.e actually exists in the view
                    if (position != RecyclerView.NO_POSITION) {
                        // Get the ride at the position, this won't work if the class is static
                        Ride ride = rides.get(position);
                        Bundle bundle = new Bundle();
                        User user = (User) ride.getDriver();
                        Log.i(TAG, String.valueOf(user));
                        bundle.putParcelable("user", user);
                        Fragment fragment = new ProfileFragment();
                        fragment.setArguments(bundle);
                        ((FragmentActivity) v.getContext()).getSupportFragmentManager().beginTransaction()
                                .replace(R.id.flContainer, fragment)
                                .addToBackStack(TAG)
                                .commit();
                    }
                }
            });
        }

        private void btnRequestListener() {
            btnRequest.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Log.i(TAG, "clicked on request ride");
                    int position = getAdapterPosition();
                    // Make sure the position is valid i.e actually exists in the view
                    if (position != RecyclerView.NO_POSITION) {
                        // Get the ride at the position, this won't work if the class is static
                        final Ride ride = rides.get(position);
                        final User currentUser = (User) ParseUser.getCurrentUser();

                        List<User> participantList = ride.getList("participants");
                        Log.i(TAG, String.valueOf(participantList.size()));
                        for(int i = 0; i < participantList.size(); i++) {
                            Log.i(TAG, "welcome part");
                            try {
                                participantList.get(i).fetch();
                                Log.i(TAG, participantList.get(i).getObjectId());
                                if(currentUser.getObjectId().equals(participantList.get(i).getObjectId())){
                                    Toast.makeText(context, "You are already a participant", Toast.LENGTH_SHORT).show();
                                    Log.i(TAG, "WELCOME HOME");
                                    return;
                                }
                            } catch (ParseException e) {
                                Log.e(TAG, "exception fetching participants", e);
                            }
                        }

                        ParseQuery<Request> query = ParseQuery.getQuery(Request.class);
                        query.whereEqualTo("ride", ride);
                        query.whereEqualTo("requester", currentUser);
                        query.include("ride");

                        // Finds the posts asynchronously
                        query.getFirstInBackground(new GetCallback<Request>() {
                            @Override
                            public void done(Request object, ParseException e) {
                                  if(object == null){
                                      Request request = new Request();
                                      request.setRequester(currentUser);
                                      request.setRide(ride);
                                      request.setDriver(ride.getDriver());
                                      save(request);
                                  }
                                  else{
                                      Toast.makeText(context, "You already requested to join this ride", Toast.LENGTH_SHORT).show();
                                  }
                            }
                        });
                    }
                }
            });
        }

        private void save(Request request) {
            request.saveInBackground(new SaveCallback() {

                @Override
                public void done(ParseException e) {
                    if (e != null) {
                        Log.e(TAG, "Error while saving", e);
                        //Toast.makeText(getContext(), "Update unsuccessful!", Toast.LENGTH_SHORT).show();
                    }
                    Log.i(TAG, "update save was successful!");
                    Toast.makeText(context, "You have requested to join this ride", Toast.LENGTH_SHORT).show();
                }
            });
        }


    }
}

File: app/src/main/java/com/example/hitchapp/adapters/RidesAdapter.kt
package com.example.hitchapp.adapters

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.hitchapp.R
import com.example.hitchapp.fragments.ProfileFragment
import com.example.hitchapp.models.Request
import com.example.hitchapp.models.Ride
import com.example.hitchapp.models.User
import com.parse.ParseException
import com.parse.ParseObject
import com.parse.ParseQuery
import com.parse.ParseUser
import java.text.SimpleDateFormat

class RidesAdapter(private val context: Context, private val rides: MutableList<Ride>) : RecyclerView.Adapter<RidesAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_ride, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val ride = rides[position]
        holder.bind(ride)
    }

    override fun getItemCount(): Int {
        return rides.size
    }

    // Clean all elements of the recycler
    fun clear() {
        rides.clear()
        notifyDataSetChanged()
    }

    // Add a list of items -- change to type used
    fun addAll(list: Collection<Ride?>) {
        rides.addAll(list as Collection<Ride>)
        notifyDataSetChanged()
    }

    fun setAll(list: List<Ride?>?) {
        clear()
        list?.let { addAll(it) }
        notifyDataSetChanged()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val tvFirstName: TextView = itemView.findViewById(R.id.tvFirstName)
        private val tvLastName: TextView = itemView.findViewById(R.id.tvLastName)
        private val ivProfilePicture: ImageView = itemView.findViewById(R.id.ivProfilePicure)
        private val tvFrom: TextView = itemView.findViewById(R.id.tvFrom)
        private val tvTo: TextView = itemView.findViewById(R.id.tvTo)
        private val tvDepartureTime: TextView = itemView.findViewById(R.id.tvDepartureTime)
        private val tvDepartureDate: TextView = itemView.findViewById(R.id.tvDepartureDate)
        private val tvPrice: TextView = itemView.findViewById(R.id.tvPrice)
        private val tvPerPerson: TextView = itemView.findViewById(R.id.tvPerPerson)
        private val tvSeatsAvailable: TextView = itemView.findViewById(R.id.tvSeatsAvailable)
        private var btnRequest: Button = itemView.findViewById(R.id.btnRequest)

        fun bind(ride: Ride) {
            val dateFor = SimpleDateFormat("MM/dd/yyyy")

            // Bind the ride data to the view elements
            val user = ride.driver as User?
            tvFirstName.text = user!!.firstName
            tvLastName.text = user.lastName
            tvFrom.text = ride.from
            tvTo.text = ride.to
            tvDepartureTime.text = ride.departureTime
            tvDepartureDate.text = dateFor.format(ride.departureDate)
            tvPrice.text = ride.price.toString()

            // fetch car data
            try {
                (ride.driver as User?)?.car?.fetch<ParseObject>()
            } catch (e: Exception) {
                Log.e(TAG, "Couldn't fetch car", e)
            }
            var seatsAvailableText = ride.seatsAvailable.toString() + "/" + (ride?.driver as User?)?.car?.carCapacity.toString()
            tvSeatsAvailable.text = seatsAvailableText
            if (ParseUser.getCurrentUser().objectId == ride.driver!!.objectId) btnRequest.visibility = View.GONE else {
                btnRequest.visibility = View.VISIBLE
            }
            if (ride.pricePerParticipant) {
                tvPerPerson.visibility = View.VISIBLE
            } else {
                tvPerPerson.visibility = View.GONE
            }
            val profile = user.profilePicture
            if (profile != null) {
                Glide.with(context)
                        .load(profile.url)
                        .fitCenter()
                        .circleCrop()
                        .into(ivProfilePicture)
            }
        }

        override fun onClick(v: View) {}

        // When someone's profile pic gets clicked you get taken to their profile
        private fun profilePicListener() {
            ivProfilePicture.setOnClickListener { v ->
                val position = adapterPosition
                // Make sure the position is valid i.e actually exists in the view
                if (position != RecyclerView.NO_POSITION) {
                    // Get the ride at the position, this won't work if the class is static
                    val ride = rides[position]
                    val bundle = Bundle()
                    val user = ride.driver as User?
                    Log.i(TAG, user.toString())
                    bundle.putParcelable("user", user)
                    val fragment: Fragment = ProfileFragment()
                    fragment.arguments = bundle
                    (v.context as FragmentActivity).supportFragmentManager.beginTransaction()
                            .replace(R.id.flContainer, fragment)
                            .addToBackStack(TAG)
                            .commit()
                }
            }
        }

        private fun btnRequestListener() {
            btnRequest.setOnClickListener(View.OnClickListener {
                Log.i(TAG, "clicked on request ride")
                val position = adapterPosition
                // Make sure the position is valid i.e actually exists in the view
                if (position != RecyclerView.NO_POSITION) {
                    // Get the ride at the position, this won't work if the class is static
                    val ride = rides[position]
                    val currentUser = ParseUser.getCurrentUser() as User
                    val participantList = ride.getList<User>("participants")
                    Log.i(TAG, participantList!!.size.toString())
                    for (i in participantList.indices) {
                        Log.i(TAG, "welcome part")
                        try {
                            participantList[i].fetch()
                            Log.i(TAG, participantList[i].objectId)
                            if (currentUser.objectId == participantList[i].objectId) {
                                Toast.makeText(context, "You are already a participant", Toast.LENGTH_SHORT).show()
                                Log.i(TAG, "WELCOME HOME")
                                return@OnClickListener
                            }
                        } catch (e: ParseException) {
                            Log.e(TAG, "exception fetching participants", e)
                        }
                    }
                    val query = ParseQuery.getQuery(Request::class.java)
                    query.whereEqualTo("ride", ride)
                    query.whereEqualTo("requester", currentUser)
                    query.include("ride")

                    // Finds the posts asynchronously
                    query.getFirstInBackground { `object`, e ->
                        if (`object` == null) {
                            val request = Request()
                            request.requester = currentUser
                            request.ride = ride
                            request.driver = ride.driver
                            save(request)
                        } else {
                            Toast.makeText(context, "You already requested to join this ride", Toast.LENGTH_SHORT).show()
                        }
                    }
                }
            })
        }

        private fun save(request: Request) {
            request.saveInBackground { e ->
                if (e != null) {
                    Log.e(TAG, "Error while saving", e)
                    //Toast.makeText(getContext(), "Update unsuccessful!", Toast.LENGTH_SHORT).show();
                }
                Log.i(TAG, "update save was successful!")
                Toast.makeText(context, "You have requested to join this ride", Toast.LENGTH_SHORT).show()
            }
        }

        init {

            // Add this as the itemView's OnClickListener
            itemView.setOnClickListener(this)

            // Listens for driver profile pic clicked
            profilePicListener()

            // Listens for when someone requests to join ride
            btnRequestListener()
        }
    }

    companion object {
        private const val TAG = "RidesAdapter"
    }

}